Images on webpage appear black

pvollan opened the pull request:
#### 557f88c94cafbd420b4aa06eeb779dd94dbc3f8e
<pre>
Images on webpage appear black
<a href="https://bugs.webkit.org/show_bug.cgi?id=254461">https://bugs.webkit.org/show_bug.cgi?id=254461</a>
rdar://107049001

Reviewed by Said Abou-Hallawa.

Images on webpage appear black with the experimental sandbox feature enabled on macOS.
This is caused by launchd being blocked in the sandbox, which prevents connections to
trustd in order to verify the code signature of image decoder plugins. This is fixed
by opening up a Mach connections to trustd before we block launchd after the WebContent
has finished launching.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/262373@main">https://commits.webkit.org/262373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2318aa29b6b0792fac67ca9c07ffe2a323a48761

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1361 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/1394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2260 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1451 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/1454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1373 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/2106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/2106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/1240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1265 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/2106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/1454 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1221 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/1237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/336 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1329 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -277,7 +278,16 @@ static void softlinkDataDetectorsFrameworks()
+    if (SandboxExtension::consumePermanently(parameters.trustdExtensionHandle)) {
+        // Open up a Mach connection to trustd by doing a code check validation.
+        // This is required since launchd will be blocked after process launch, which prevents new Mach connections to be created.
+        auto bundleURL = adoptCF(CFURLCreateWithFileSystemPath(kCFAllocatorDefault, CFSTR("/System/Library/Video/Plug-Ins/AppleGVAHEVCDecoder.bundle"), kCFURLPOSIXPathStyle, false));
```
This seems very fragile and not future-proof. What if that path changes, or some other plugin starts to be used for some image types?

pvollan replied:
That is a good point, Simon. In this case, the actual path used isn't important, since the purpose is to just open up a connection to trustd. I will add a comment and a FIXME about this.

Thanks for reviewing!

geoffreygaren replied:
If AppleGVAHEVCDecoder.bundle is not present on the system will trustd still launch to verify its signature?

pvollan replied:
> If AppleGVAHEVCDecoder.bundle is not present on the system will trustd still launch to verify its signature?

Good point! I will have to check that manually by specifying a non-existent path.

Thanks for reviewing!

pvollan replied:
> > If AppleGVAHEVCDecoder.bundle is not present on the system will trustd still launch to verify its signature?
> 
> Good point! I will have to check that manually by specifying a non-existent path.
> 
> Thanks for reviewing!

Good catch! I have changed the bundle URL to address this.

shallawa commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -277,7 +278,18 @@ static void softlinkDataDetectorsFrameworks()
+        auto bundleURL = adoptCF(CFURLCreateWithFileSystemPath(kCFAllocatorDefault, CFSTR("/System/Library/Video/Plug-Ins/AppleGVAHEVCDecoder.bundle"), kCFURLPOSIXPathStyle, false));
+        SecStaticCodeRef pluginCode = nullptr;
+        SecStaticCodeCreateWithPath(bundleURL.get(), kSecCSDefaultFlags, &pluginCode);
+        SecStaticCodeCheckValidity(pluginCode, kSecCSDoNotValidateResources, nullptr);
```
Is it okay and necessary to call SecStaticCodeCheckValidity() if pluginCode is nullptr?

pvollan replied:
> Is it okay and necessary to call SecStaticCodeCheckValidity() if pluginCode is nullptr?

That is a good point. I have addressed this.

Thanks for reviewing!

shallawa commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -277,7 +278,19 @@ static void softlinkDataDetectorsFrameworks()
+    if (SandboxExtension::consumePermanently(parameters.trustdExtensionHandle)) {
+        // Open up a Mach connection to trustd by doing a code check validation.
+        // This is required since launchd will be blocked after process launch, which prevents new Mach connections to be created.
+        // The actual path is not important in this case, since we just want to open a Mach connection.
```
Are we sure that opening a Mach connection will not violate the sandbox itself? It is just weird that we apply the sandbox but we have to allow frameworks outside it such as trustd and launchd.

pvollan replied:
This will not cause a sandbox violation, since we allow connections to trustd with the extension, which is consumed here. The issue is that opening up connections to trustd will fail after the WebContent process has finished launching, since we then block launchd, which is required to set up the connection to trustd. We fix this by eagerly creating a connection to. trustd here.

Thanks for reviewing!

pvollan committed:
Images on webpage appear black
https://bugs.webkit.org/show_bug.cgi?id=254461
rdar://107049001

Reviewed by Said Abou-Hallawa.

Images on webpage appear black with the experimental sandbox feature enabled on macOS.
This is caused by launchd being blocked in the sandbox, which prevents connections to
trustd in order to verify the code signature of image decoder plugins. This is fixed
by opening up a Mach connections to trustd before we block launchd after the WebContent
has finished launching.

* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: https://commits.webkit.org/262373@main
